### ISSUE TITLE
Flexible routes

### ISSUE THREAD

**Issue by benbovy** [xarray-contrib/xpublish#25]
I would be nice if xpublish could have some kind of mechanism to easily set which routes will be added to the application.

I guess it is technically possible as `RestAccessor` publicly exposes the FastAPI application with the `app` property, but it is less clear to me how we could remove some existing routes (e.g., in case we just want to publish dataset info but not zarr-like data access) or define some new routes outside `RestAccessor` and still have access to the xarray Dataset object from within the path operation functions.

Instead, I'm thinking about something like this:

```python
Dataset.rest(routers=[...]).serve()
```

where `routers` accepts a sequence of `fastapi.APIRouter` instances.

This would be a good opportunity to refactor the current routes into multiple (fine-grained) routers, e.g., for basic things (root path, versions), dataset info and zarr-like data access.

Ideally, we should also be able to easily inject dependencies and/or middleware to any router, e.g., to require authentication for some part of the API (e.g., zarr-like data access).

I started a prototype in [this gist](https://gist.github.com/benbovy/f687e1697e18e39e4924872cd83a60e9).

The basic idea is to wrap `fastapi.APIRouter` so that we can inject a fastapi dependency for accessing the dataset from within the router's path operation functions. As a bonus, this prototype allows routers to be reused for serving either one single dataset or a collection of datasets! That could be a nice solution for #23.

The prototype works well. I'm still not super happy with it, though. `ExampleRouter` could look better IMO.

**Comment by benbovy** [comment]
> The prototype works well. I'm still not super happy with it, though.

I think a much cleaner approach would be to use FastAPI's [dependency overriding](https://fastapi.tiangolo.com/advanced/testing-dependencies/). I updated the [gist](https://gist.github.com/benbovy/f687e1697e18e39e4924872cd83a60e9#file-flexible_routes_alt-py). See also #27.

The only drawback I've found with this approach is that the auto-generated API documentation is not updated properly when the signature of the overridden dependency changes. In the gist above, `single_app` endpoints still show an optional `dataset_id` query parameter that . That's not a big deal, though, we can just ignore that parameter in this case. Maybe this could be fixed upstream.
